### PR TITLE
Parallelize GitHub Action jobs in CI

### PR DESCRIPTION
**Pull request by torrocus**
## Pull Request Summary

CI tests take a long time because they run different cases one after the other. If something fails, no further tests will be run. ~CI tests have not been passed for some time now. Fast tests didn't pass on CI for MacOS. This change is intended to help us solve this problem more easily.~

Changes proposed in this pull request:
* split a long-running tests job into smaller jobs that can be parallelized

## Feedback

The total CI execution time is the same, but the waiting time for the test results is shorter.

### Before

Screenshot before

### After

Screenshot after

### CODE REVIEW

**Comment by torrocus** [comment]
@pkuczynski what do you think about this change?

File: .github/workflows/specs.yml
Comment (pkuczynski) [inline .github/workflows/specs.yml:85]: Isnt this one for trufflefurby?

**Comment by pkuczynski** [comment]
It make sense to run things in parallel if it speeds up the tests, that's for sure. Just the code duplication is aweful. Can we somehow avoid this?

@eregon wdyt?

**Comment by eregon** [comment]
What about using a 2-dimensional matrix for that? That should address the duplication and keep things concise.

**Comment by eregon** [comment]
> CI tests have not been passed for some time now.

Only on macOS which is a PITA. They work fine on Linux AFAIK.

File: .github/workflows/specs.yml
Comment (eregon) [inline .github/workflows/specs.yml:8]: I don't think we should cancel CI jobs

Comment (torrocus) [reply .github/workflows/specs.yml:8]: Let me explain it. Canceling is only when we push new changes to the same branch. This allows you to save running time on GitHub if you make frequent changes. This does not affect CI if work is carried out on different branches.

Comment (eregon) [reply .github/workflows/specs.yml:8]: I know what it does, but for instance it could cancel a CI run on master if 2 PRs are merged quickly one after another, and that seems bad (could make it harder to find a faulty change) + there is not really enough CI load on this repo to need to cancel anything.

Comment (torrocus) [reply .github/workflows/specs.yml]: We can exclude cancel for branch master. WDYT?

Comment (eregon) [reply .github/workflows/specs.yml]: But then that prevents to run CI jobs on different commits on master in parallel, which is unnecessary. Please remove this section entirely, there is no need to change this.

File: .github/workflows/specs.yml
Comment (torrocus) [reply .github/workflows/specs.yml:85]: I don't want to guess, so let's check it out.

Comment (torrocus) [reply .github/workflows/specs.yml:85]: The [tests didn't pass](https://github.com/rvm/rvm/actions/runs/8940912966), so no.

**Comment by torrocus** [comment]
> @pkuczynski said:
> It make sense to run things in parallel if it speeds up the tests, that's for sure. Just the code duplication is aweful. Can we somehow avoid this?

I don't like these duplications either. Unfortunately, every job must perform similar activities:
- git checkout repo
- install ruby via rvm
- install tf gem
- run selected tests

At the moment I have no idea how to do it better, although I would love to share these parts of the job.


> @eregon said:
> What about using a 2-dimensional matrix for that? That should address the duplication and keep things concise.

If you mean strategies (matrix for OS), it won't change anything. To introduce parallelism, we need separate jobs. Anyway, where we run jobs on different OSs, I used matrix. I only removed where it was conditional.

**Comment by eregon** [comment]
Some general feedback: I think 9min vs 14min doesn't make a huge difference, in either case no one will wait for that actively. And then of course it will redo some work multiple times which is rather wasteful (e.g. compile ruby 2.7.7).
But having isolation between these groups of tests is valuable, especially given the existing issues between them, so this seems worth it.

**Comment by eregon** [comment]
> If you mean strategies (matrix for OS), it won't change anything. To introduce parallelism, we need separate jobs. Anyway, where we run jobs on different OSs, I used matrix. I only removed where it was conditional.

Use a matrix e.g. like here: https://github.com/ruby/spec/blob/1afa2eab1cab5df6c5706fc5b7e92756b1db4f07/.github/workflows/ci.yml#L10-L12
Those jobs all run in parallel.
And that solves the duplication, at the cost of some existing `if` conditionals for tests which don't work on macOS yet (but should eventually).

You can even avoid those `if:` with an explicit include list like https://github.com/ruby/spec/blob/1afa2eab1cab5df6c5706fc5b7e92756b1db4f07/.github/workflows/ci.yml#L14-L17

File: .github/workflows/specs.yml
Comment (eregon) [reply .github/workflows/specs.yml:85]: I think it's more the opposite: `truffleruby_comment_test.sh` might not need `2.7.7` to be installed. But it feels something out of scope and might not hold either, so I wouldn't change that in this PR.

**Comment by torrocus** [comment]
> Some general feedback: I think 9min vs 14min doesn't make a huge difference, in either case no one will wait for that actively. And then of course it will redo some work multiple times which is rather wasteful (e.g. compile ruby 2.7.7). But having isolation between these groups of tests is valuable, especially given the existing issues between them, so this seems worth it.

Yes, separation is a big advantage. If we add caching in the future, we should gain more time.

**Comment by torrocus** [comment]
> > If you mean strategies (matrix for OS), it won't change anything. To introduce parallelism, we need separate jobs. Anyway, where we run jobs on different OSs, I used matrix. I only removed where it was conditional.
> 
> Use a matrix e.g. like here: https://github.com/ruby/spec/blob/1afa2eab1cab5df6c5706fc5b7e92756b1db4f07/.github/workflows/ci.yml#L10-L12 (the keys would be `os` and `test` instead) Those jobs all run in parallel. And that solves the duplication, at the cost of some existing `if` conditionals for tests which don't work on macOS yet (but should eventually).
> 
> You can even avoid those `if:` with an explicit include list like https://github.com/ruby/spec/blob/1afa2eab1cab5df6c5706fc5b7e92756b1db4f07/.github/workflows/ci.yml#L14-L17

Interesting idea. I'll see what I can do with it.

**Comment by torrocus** [comment]
@pkuczynski @eregon I made changes and it looks better. What do you think?

File: .github/workflows/specs.yml
Comment (eregon) [inline .github/workflows/specs.yml:21]: Could you preserve the comments for these 2 jobs (as to why they are not run on macOS)?

File: .github/workflows/specs.yml
Comment (eregon) [inline .github/workflows/specs.yml]: Could you remove this?

File: .github/workflows/specs.yml
Comment (torrocus) [reply .github/workflows/specs.yml]: I feel like removing will hurt the project, but I'll adapt.

**Review by eregon — COMMENTED** [review]
Looks great now, 2 more comments

File: .github/workflows/specs.yml
Comment (eregon) [reply .github/workflows/specs.yml]: Because CI would start with a delay? I've never observed it on rvm/rvm, so I think it's a not an issue in practice (the default numbers of runners is pretty high).
And for the scenario I mentioned for master it's actually worse, adding a constraint which is not needed.
Another downside is e.g. when working on CI one might push more changes but still be interested in older in-progress CI jobs (e.g. I still want to see macOS CI results when touching common logic). This would cancel them and make it very annoying for those cases, which IMO outweighs any possible gain from this.

**Review by eregon — APPROVED** [review]

File: .github/workflows/specs.yml
Comment (torrocus) [reply .github/workflows/specs.yml]: The way I see it is that if I make changes, I would like to minimize the time in which I receive information that the tests have passed and I don't want to wait until the entire CI queue is completed. I completely agree that the master branch should be unique in this case. But maybe it won't be such a big problem. If it hurts, we can consider the proposed changes.

File: .github/workflows/specs.yml
Comment (eregon) [inline .github/workflows/specs.yml:19]: This comment applies to `tests/fast/*`, not `tests/long/truffleruby_comment_test.sh`.
But I think best to remove it, it's not relevant anymore since they run in separate jobs.

**Review by eregon — DISMISSED** [review]

File: .github/workflows/specs.yml
Comment (torrocus) [reply .github/workflows/specs.yml:19]: Right, I missed that. Well caught.

**Review by eregon — APPROVED** [review]

**Comment by eregon** [comment]
@pkuczynski Could you merge? (I cannot)
And also adapt the Required status checks to the new job names?

**Comment by torrocus** [comment]
@pkuczynski Since we have changed the CI process, we also need to change the settings in -> Settings -> Branches -> Branch protection rules. Then select Edit for the "master" branch. And then in the "Require status checks to pass before merging" option, remove the status checks "tests" and add "CI/tests".
Otherwise, only people with force push permission can merge ("tests" check status is waiting for report over and over again).

**Comment by pkuczynski** [comment]
Sure I can :) Give me few mins.

**Review by pkuczynski — APPROVED** [review]

**Comment by pkuczynski** [comment]
Thank you @torrocus for your contribution

**Comment by eregon** [comment]
> MacOS tests are taking ridicilous long. Maybe we could check why...

Looking at https://github.com/rvm/rvm/actions/runs/8945080363/usage and specific logs for test/fast/*, it looks like on Linux it's mostly (only?) prebuilt Ruby binaries being used, while on macOS every ruby installed seems compiled from source, which takes a while.

One thing to improve might be to always use the same Ruby version for all tests and avoid uninstalling it.
And obviously if there could be prebuilt Rubies on macOS that'd be even better.

**Comment by pkuczynski** [comment]
That make sense. We had issues with building movable ruby in the past for macOS :(

**Pull request merged**
